### PR TITLE
gzip json responses if headers say its okay

### PR DESCRIPTION
**Pull request by CannonLock**
Checks for accept-encoding:gzip and if it is there then gzip json responses to save some bandwidth

### CODE REVIEW

File: src/app.py [hunk 1]
```diff
@@ -884,6 +884,20 @@ def _get_authorized():
     # If it gets here, then it is not authorized
     return default_authorized
 
+def make_json_response(json_response: dict) -> flask.Response:
+    """Trys to gzip json response"""
+
+    if 'gzip' in flask.request.headers.get("accept-encoding", ""):
+        content = gzip.compress(json.dumps(json_response).encode('utf8'), 5)
+        response = flask.make_response(content)
+        response.headers['Content-length'] = len(content)
```
Comment (matyasselmeci) [inline src/app.py:893]:
```suggestion
        content = gzip.compress(json.dumps(json_response).encode('utf8', errors='replace'), 5)
        response = flask.make_response(content)
        response.headers['Content-Length'] = len(content)
```

File: src/tests/test_api.py [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import re
-import flask
-import pytest
+import gzip
+import json
 import urllib.parse
```
Comment (matyasselmeci) [inline src/tests/test_api.py:4]:
```suggestion
from typing import Generator
import urllib.parse
```

File: src/tests/test_api.py [hunk 3]
```diff
@@ -59,25 +62,36 @@
...
+    '/miscresource/json',
+    '/vosummary/json',
+    '/resources/stashcache-files',
+    '/stashcache/namespaces'
+]
+
 
 @pytest.fixture
-def client():
+def client() -> FlaskClient:
```
Comment (matyasselmeci) [inline src/tests/test_api.py:78]:
```suggestion
def client() -> Generator[FlaskClient]:
```
since you're using a `yield` statement

**Review by matyasselmeci — APPROVED** [review]
Some tweaks but otherwise LGTM. Pre-approving.

**Comment by CannonLock** [comment]
@djw8605 Same. That said its not often I get to implement 10x performance upgrades so I am happy they give me the opportunity. :)

File: src/tests/test_api.py [hunk 4]
```diff
@@ -59,25 +63,36 @@
...
+    '/miscresource/json',
+    '/vosummary/json',
+    '/resources/stashcache-files',
+    '/stashcache/namespaces'
+]
+
 
 @pytest.fixture
-def client():
+def client() -> Generator[FlaskClient]:
```
Comment (matyasselmeci) [inline src/tests/test_api.py:79]:
```suggestion
def client() -> Generator[FlaskClient, None, None]:
```
Sorry, apparently the syntax for generators is
```
Generator[yield type, send type, return type]
```

**Comment by jthiltges** [comment]
This might be an alternative method: https://github.com/opensciencegrid/topology/commit/ba9b9fe6b2ab7156e46b3beee9939b6c12e1f709

**Comment by CannonLock** [comment]
@jthiltges Does that zip everything, header independent? I have some things that will bork if they get back a zipped response.

**Comment by jthiltges** [comment]
> @jthiltges Does that zip everything, header independent? I have some things that will bork if they get back a zipped response.

Good question! Quickly glancing at the docs, I'd say yes: [AddOutputFilterByType](https://httpd.apache.org/docs/2.4/mod/mod_filter.html#addoutputfilterbytype) uses [mod_deflate](https://httpd.apache.org/docs/2.4/mod/mod_deflate.html#enable), which sounds like it does the right thing in respecting accept-encoding (unless forced otherwise).

**Comment by CannonLock** [comment]
Closing in favor of doing this on the apache level -> #3133

**Pull request closed without merging**